internal/jobqueue/dbjobqueue/dbjobqueue.go

```diff
@@ -0,0 +1,390 @@
+}
+
+func (q *dbJobQueue) Enqueue(jobType string, args interface{}, dependencies []uuid.UUID) (uuid.UUID, error) {
+	conn, err := q.pool.Acquire(context.Background())
+	if err != nil {
+		return uuid.Nil, fmt.Errorf("error connecting to database: %v", err)
+	}
+	defer conn.Release()
+
+	tx, err := conn.Begin(context.Background())
+	if err != nil {
+		return uuid.Nil, fmt.Errorf("error starting database transaction: %v", err)
+	}
+	defer func() { _ = tx.Rollback(context.Background()) }()
+
+	id := uuid.New()
+	_, err = conn.Exec(context.Background(), sqlEnqueue, id, jobType, args)
+	if err != nil {
+		return uuid.Nil, fmt.Errorf("error enqueuing job: %v", err)
+	}
+
+	for _, d := range dependencies {
+		_, err = conn.Exec(context.Background(), sqlInsertDependency, id, d)
+		if err != nil {
+			return uuid.Nil, fmt.Errorf("error inserting dependency: %v", err)
+		}
+	}
+
+	_, err = conn.Exec(context.Background(), sqlNotify)
+	if err != nil {
+		return uuid.Nil, fmt.Errorf("error notifying jobs channel: %v", err)
+	}
+
+	err = tx.Commit(context.Background())
+	if err != nil {
+		return uuid.Nil, fmt.Errorf("unable to commit database transaction: %v", err)
+	}
+
+	return id, nil
+}
+
+func (q *dbJobQueue) Dequeue(ctx context.Context, jobTypes []string) (uuid.UUID, uuid.UUID, []uuid.UUID, string, json.RawMessage, error) {
+	// Return early if the context is already canceled.
+	if err := ctx.Err(); err != nil {
+		return uuid.Nil, uuid.Nil, nil, "", nil, err
+	}
+
+	conn, err := q.pool.Acquire(ctx)
+	if err != nil {
+		return uuid.Nil, uuid.Nil, nil, "", nil, fmt.Errorf("error connecting to database: %v", err)
+	}
+	defer func() {
+		_, _ = conn.Exec(ctx, sqlUnlisten) // ignore errors when cleaning up
+		conn.Release()
+	}()
+
+	_, err = conn.Exec(ctx, sqlListen)
+	if err != nil {
+		return uuid.Nil, uuid.Nil, nil, "", nil, fmt.Errorf("error listening on jobs channel: %v", err)
+	}
+
+	var id uuid.UUID
+	var jobType string
+	var args json.RawMessage
+	token := uuid.New()
+	for {
+		err = conn.QueryRow(ctx, sqlDequeue, token, jobTypes).Scan(&id, &token, &jobType, &args)
+		if err == nil {
+			break
+		}
+		if err != nil && !errors.As(err, &pgx.ErrNoRows) {
+			return uuid.Nil, uuid.Nil, nil, "", nil, fmt.Errorf("error dequeuing job: %v", err)
+		}
+		_, err = conn.Conn().WaitForNotification(ctx)
+		if err != nil {
+			return uuid.Nil, uuid.Nil, nil, "", nil, fmt.Errorf("error waiting for notification on jobs channel: %v", err)
+		}
+	}
+
+	// insert heartbeat
+	_, err = conn.Exec(ctx, sqlInsertHeartbeat, token, id)
+	if err != nil {
+		return uuid.Nil, uuid.Nil, nil, "", nil, fmt.Errorf("error inserting the job's heartbeat: %v", err)
+	}
+
+	dependencies, err := q.jobDependencies(ctx, conn, id)
+	if err != nil {
+		return uuid.Nil, uuid.Nil, nil, "", nil, fmt.Errorf("error querying the job's dependencies: %v", err)
+	}
+
+	return id, token, dependencies, jobType, args, nil
+}
+
+func (q *dbJobQueue) FinishJob(id uuid.UUID, result interface{}) error {
```

Comment:
I think this method should send a notification because finishing one job can trigger another one.